### PR TITLE
Support unified memory

### PR DESCRIPTION
**Pull request by okuta**
This PR is to avoid GPU memory limit.

### CODE REVIEW

File: cupy/cuda/memory.pxd [hunk 1]
```diff
@@ -1,17 +1,23 @@
-from cupy.cuda cimport device
+from cupy.cuda cimport device as _device
```
Comment (unnonouno) [inline cupy/cuda/memory.pxd]: How about using `device_mod`?

File: cupy/cuda/memory.pyx [hunk 2]
```diff
@@ -8,11 +8,15 @@ import weakref
 import six
 
 from cupy.cuda import runtime
+from cupy.cuda import stream
 
-from cupy.cuda cimport device
+from cupy.cuda cimport device as _device
```
Comment (unnonouno) [inline cupy/cuda/memory.pyx]: ditto

File: cupy/cuda/memory.pyx [hunk 3]
```diff
@@ -8,11 +8,15 @@ import weakref
...
 from cupy.cuda cimport runtime
 
 
+_cuda_version = None
```
Comment (unnonouno) [inline cupy/cuda/memory.pyx]: How about calling runtimeGetVersion here?

File: cupy/cuda/memory.pyx [hunk 4]
```diff
@@ -42,6 +46,55 @@ cdef class Memory:
...
+        """ Prefetch.
+
+        Args:
+            stream (cupy.cuda.Stream): Stream
+
+        """
+        global _cuda_version
+        if _cuda_version is None:
+            _cuda_version = runtime.runtimeGetVersion()
+        if _cuda_version >= 8000 and int(self.device.compute_capability) >= 60:
```
Comment (unnonouno) [inline cupy/cuda/memory.pyx]: Do we need to raise Exception when unified memory is not supported?

File: cupy/cuda/memory.pyx [hunk 5]
```diff
@@ -42,6 +46,55 @@ cdef class Memory:
...
+
+        Args:
+            advics (int): Advise to be applied for this memory.
+            device (cupy.cuda.Device): Device to apply the advice for.
+
+        """
+        global _cuda_version
+        if _cuda_version is None:
+            _cuda_version = runtime.runtimeGetVersion()
+        if _cuda_version >= 8000 and int(self.device.compute_capability) >= 60:
```
Comment (unnonouno) [inline cupy/cuda/memory.pyx]: ditto

File: cupy/cuda/memory.pyx [hunk 6]
```diff
@@ -261,6 +314,14 @@ cpdef MemoryPointer _malloc(Py_ssize_t size):
     return MemoryPointer(mem, 0)
 
 
+cpdef MemoryPointer _mallocManaged(Py_ssize_t size):
+    mem = ManagedMemory(size)
```
Comment (unnonouno) [inline cupy/cuda/memory.pyx]: `cedf ManagedMemory mem` ?

File: cupy/cuda/memory.pyx (hunk 4 above)
Comment (okuta) [inline cupy/cuda/memory.pyx]: This method is using for only performance up.
It is better to do nothing on old CUDA environment.

File: cupy/cuda/memory.pyx [hunk 7]
```diff
@@ -262,6 +309,14 @@ cpdef MemoryPointer _malloc(Py_ssize_t size):
     return MemoryPointer(mem, 0)
 
 
+cpdef MemoryPointer _mallocManaged(Py_ssize_t size):
```
Comment (unnonouno) [inline cupy/cuda/memory.pyx]: How about using snake case?

File: cupy/cuda/memory.pyx [hunk 8]
```diff
@@ -43,6 +47,49 @@ cdef class Memory:
...
+            self.ptr = runtime.mallocManaged(size)
+
+    cpdef prefetch(self, stream):
+        """ Prefetch.
+
+        Args:
+            stream (cupy.cuda.Stream): Stream
+
+        """
+        if _cuda_version >= 8000 and int(self.device.compute_capability) >= 60:
```
Comment (unnonouno) [inline cupy/cuda/memory.pyx]: How about checking cuda version in __init__?

File: cupy/cuda/memory.pyx (hunk 7 above)
Comment (unnonouno) [inline cupy/cuda/memory.pyx]: Do we need to check cuda version?

File: examples/imagenet/train_imagenet.py [hunk 9]
```diff
@@ -23,6 +23,11 @@
 import googlenetbn
 import nin
 
+import cupy
+
+memory_pool = cupy.cuda.MemoryPool(cupy.cuda.memory._mallocManaged)
```
Comment (unnonouno) [inline examples/imagenet/train_imagenet.py]: Is _mallocManaged public?

**Comment by okuta** [comment]
@unnonouno I removed some code. Please review this.

**Comment by unnonouno** [comment]
Check my comments.

File: cupy/cuda/memory.pyx [hunk 10]
```diff
@@ -43,6 +47,52 @@ cdef class Memory:
...
+
+    cpdef advise(self, int advise, device_mod.Device device):
+        """ Advise about the usage of this memory.
+
+        Args:
+            advics (int): Advise to be applied for this memory.
+            device (cupy.cuda.Device): Device to apply the advice for.
+
+        """
+        if self._support_prefetch_advise:
```
Comment (unnonouno) [inline cupy/cuda/memory.pyx]: Do we need to show warning? Users cannot know that this object ignores this method.

File: cupy/cuda/memory.pyx [hunk 11]
```diff
@@ -43,6 +47,52 @@ cdef class Memory:
         return self.ptr
 
 
+cdef class ManagedMemory(Memory):
+
+    """Memory allocation on a CUDA device.
```
Comment (unnonouno) [inline cupy/cuda/memory.pyx]: Is this class for unified memory? It is difficult to understand its role.

File: cupy/cuda/memory.pyx [hunk 12]
```diff
@@ -43,6 +47,57 @@ cdef class Memory:
...
+        self.ptr = 0
+        if size > 0:
+            self.device = device_mod.Device()
+            self.ptr = runtime.mallocManaged(size)
+            self._support_prefetch_advise = (
+                _cuda_version >= 8000 and
+                int(self.device.compute_capability) >= 60)
+
+    cpdef prefetch(self, stream):
+        """ Prefetch.
```
Comment (unnonouno) [inline cupy/cuda/memory.pyx]: "Prefetch memory." ?

File: cupy/cuda/memory.pyx [hunk 13]
```diff
@@ -262,6 +317,19 @@ cpdef MemoryPointer _malloc(Py_ssize_t size):
     return MemoryPointer(mem, 0)
 
 
+cpdef MemoryPointer malloc_managed(Py_ssize_t size):
+    """Allocate managed memory
```
Comment (unnonouno) [inline cupy/cuda/memory.pyx]: Add period.

File: cupy/cuda/memory.pyx [hunk 14]
```diff
@@ -43,6 +47,57 @@ cdef class Memory:
...
+                int(self.device.compute_capability) >= 60)
+
+    cpdef prefetch(self, stream):
+        """ Prefetch.
+
+        Args:
+            stream (cupy.cuda.Stream): Stream
+
+        """
+        if not self._support_prefetch_advise:
```
Comment (unnonouno) [inline cupy/cuda/memory.pyx]: How about adding argument to control behavior on unsupported situation? Such as `error='ignoer'`.

**Comment by unnonouno** [comment]
Please sent it cupy/cupy

**Comment by okuta** [comment]
I send PR cupy/cupy#163.

**Pull request closed without merging**